src/macaron/slsa_analyzer/checks/infer_artifact_pipeline_check.py

```diff
@@ -93,35 +116,55 @@ def run_check(self, ctx: AnalyzeContext) -> CheckResultData:
         CheckResultData
             The result type of the check.
         """
-        # This check requires the build_as_code check to pass and a repository to be available.
+        # This check requires a repository to be available.
         if not ctx.component.repository:
             return CheckResultData(result_tables=[], result_type=CheckResultType.FAILED)
 
         # Look for the artifact in the corresponding registry and find the publish timestamp.
         artifact_published_date = None
-        package_registry_info_entries = ctx.dynamic_data["package_registries"]
-        for package_registry_info_entry in package_registry_info_entries:
-            match package_registry_info_entry:
-                # TODO: add package registries for other ecosystems.
-                case PackageRegistryInfo(
-                    build_tool=Gradle() | Maven(),
-                    package_registry=MavenCentralRegistry() as mvn_central_registry,
-                ):
-                    group_id = ctx.component.namespace
-                    artifact_id = ctx.component.name
-                    version = ctx.component.version
-                    try:
-                        artifact_published_date = mvn_central_registry.find_publish_timestamp(
-                            group_id, artifact_id, version
-                        )
-                    except InvalidHTTPResponseError as error:
-                        logger.debug(error)
+        for registry_info in ctx.dynamic_data["package_registries"]:
+            if registry_info.build_tool.purl_type == ctx.component.type:
+                try:
+                    artifact_published_date = registry_info.package_registry.find_publish_timestamp(ctx.component.purl)
+                    break
+                except InvalidHTTPResponseError as error:
+                    logger.debug(error)
+                except NotImplementedError:
+                    continue
```

Comment:
I think catching `NotImplementedError` here has a strong indication that the interface of `find_publish_timestamp` from subtype of `PackageRegistry` will raise `NotImplementedError`. I believe this is not the case because only Jfrog Maven registry will raise `NotImplementedError`. A better way to communicate this would be to explicitly skip running `find_publish_timestamp` if `registry_info` is of type `JFrogMavenRegistry`.
```
        for registry_info in ctx.dynamic_data["package_registries"]:
            if isinstance(registry_info.package_registry, JfrogMavenRegistry):
                # We currently don't support this.
                continue 
            if registry_info.build_tool.purl_type == ctx.component.type:
                try:
                    artifact_published_date = registry_info.package_registry.find_publish_timestamp(ctx.component.purl)
                    break
                except InvalidHTTPResponseError as error:
                    logger.debug(error)
```
What do you think?

Reply:
I prefer to catch `NotImplementedError` directly, rather than checking for specific subclass names. This approach is more general and avoids unnecessary dependencies on subclassing details.